lib/db/schema/patch-095-096.sql

```diff
@@ -0,0 +1,298 @@
+
+
+CREATE PROCEDURE `deleteAccount_16` (
+  IN `uidArg` BINARY(16)
+)
+BEGIN
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  START TRANSACTION;
+
+  DELETE FROM sessionTokens WHERE uid = uidArg;
+  DELETE FROM keyFetchTokens WHERE uid = uidArg;
+  DELETE FROM accountResetTokens WHERE uid = uidArg;
+  DELETE FROM passwordChangeTokens WHERE uid = uidArg;
+  DELETE FROM passwordForgotTokens WHERE uid = uidArg;
+  DELETE FROM accounts WHERE uid = uidArg;
+  DELETE devices, deviceCommands FROM devices LEFT JOIN deviceCommands
+    ON (deviceCommands.uid = devices.uid AND deviceCommands.deviceId = devices.id)
+    WHERE devices.uid = uidArg;
+  DELETE FROM unverifiedTokens WHERE uid = uidArg;
+  DELETE FROM unblockCodes WHERE uid = uidArg;
+  DELETE FROM emails WHERE uid = uidArg;
+  DELETE FROM signinCodes WHERE uid = uidArg;
+  DELETE FROM totp WHERE uid = uidArg;
+  DELETE FROM recoveryKeys WHERE uid = uidArg;
+  DELETE FROM recoveryCodes WHERE uid = uidArg;
+  DELETE FROM securityEvents WHERE uid = uidArg;
+
+  COMMIT;
+END;
+
+CREATE PROCEDURE `resetAccount_12` (
+  IN `uidArg` BINARY(16),
+  IN `verifyHashArg` BINARY(32),
+  IN `authSaltArg` BINARY(32),
+  IN `wrapWrapKbArg` BINARY(32),
+  IN `verifierSetAtArg` BIGINT UNSIGNED,
+  IN `verifierVersionArg` TINYINT UNSIGNED
+)
+BEGIN
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  START TRANSACTION;
+
+  DELETE FROM sessionTokens WHERE uid = uidArg;
+  DELETE FROM keyFetchTokens WHERE uid = uidArg;
+  DELETE FROM accountResetTokens WHERE uid = uidArg;
+  DELETE FROM passwordChangeTokens WHERE uid = uidArg;
+  DELETE FROM passwordForgotTokens WHERE uid = uidArg;
+  DELETE devices, deviceCommands FROM devices LEFT JOIN deviceCommands
+    ON (deviceCommands.uid = devices.uid AND deviceCommands.deviceId = devices.id)
+    WHERE devices.uid = uidArg;  DELETE FROM unverifiedTokens WHERE uid = uidArg;
+  UPDATE accounts
+  SET
+    verifyHash = verifyHashArg,
+    authSalt = authSaltArg,
+    wrapWrapKb = wrapWrapKbArg,
+    verifierSetAt = verifierSetAtArg,
+    verifierVersion = verifierVersionArg,
+    profileChangedAt = verifierSetAtArg
+  WHERE uid = uidArg;
+
+  COMMIT;
+END;
+
+CREATE PROCEDURE `deleteSessionToken_4` (
+  IN `tokenIdArg` BINARY(32)
+)
+BEGIN
+  DELETE sessionTokens, unverifiedTokens, devices, deviceCommands
+  FROM sessionTokens
+  LEFT JOIN unverifiedTokens
+    ON sessionTokens.tokenId = unverifiedTokens.tokenId
+  LEFT JOIN devices
+    ON (sessionTokens.uid = devices.uid AND sessionTokens.tokenId = devices.sessionTokenId)
+  LEFT JOIN deviceCommands
+    ON (deviceCommands.uid = devices.uid AND deviceCommands.deviceId = devices.id)
+  WHERE
+    sessionTokens.tokenId = tokenIdArg;
+END;
+
+-- Now we can actually add the new columns!
+
+ALTER TABLE devices
+  ADD COLUMN refreshTokenId BINARY(32) DEFAULT NULL,
+  ADD CONSTRAINT UQ_devices_refreshTokenId UNIQUE (uid, refreshTokenId),
```

Comment:
I asked above and will here too, is it possible for a device to have neither a refreshTokenId nor a sessionTokenId, and if so, what does that mean? If devices like that shouldn't exist, can we add a constraint that enforces one or the other?

Reply:
It is possible, and IIRC the queries go out of their way to ignore such records (referred to as "zombie devices" in some of the comments) so I don't think they cause much harm.  We could add a `CHECK` constraint to guard against it if you feel strongly about it.  (Although it's kind of against the prevailing conventions in this repo that don't use MySQL's integrity checks).